editor/src/components/canvas/grid-panels-container.tsx

```diff
@@ -0,0 +1,138 @@
+import React from 'react'
+import { accumulate, arrayAccumulate } from '../../core/shared/array-utils'
+import { GridPanel } from './grid-panel'
+import { ColumnDragTargets } from './grid-panels-drag-targets'
+import type { LayoutUpdate, StoredPanel } from './grid-panels-state'
+import {
+  GridHorizontalExtraPadding,
+  GridMenuDefaultPanels,
+  GridMenuWidth,
+  GridPaneWidth,
+  GridPanelHorizontalGapHalf,
+  GridPanelVerticalGapHalf,
+  GridVerticalExtraPadding,
+  NumberOfColumns,
+  storedLayoutToResolvedPanels,
+  updateLayout,
+  wrapAroundColIndex,
+} from './grid-panels-state'
+import { CanvasFloatingToolbars } from './camvas-floating-toolbars'
+
+export const GridPanelsContainer = React.memo(() => {
+  const [panelState, setPanelState] = React.useState(GridMenuDefaultPanels)
+
+  const orderedPanels = React.useMemo(() => {
+    return storedLayoutToResolvedPanels(panelState)
+  }, [panelState])
+
+  const nonEmptyColumns = React.useMemo(() => {
+    return Array.from(
+      accumulate(new Set<number>(), (acc: Set<number>) => {
+        // we always include the first and last columns
+        acc.add(0)
+        acc.add(NumberOfColumns - 1)
+
+        panelState.forEach((column, colIndex) => {
+          if (column.length > 0) {
+            acc.add(wrapAroundColIndex(colIndex))
+          }
+        })
+      }),
+    )
+  }, [panelState])
+
+  const onDrop = React.useCallback(
+    (itemToMove: StoredPanel, newPosition: LayoutUpdate) => {
+      setPanelState((panels) => updateLayout(panels, itemToMove, newPosition))
+    },
+    [setPanelState],
+  )
+
+  const canDrop = React.useCallback(
+    (itemToMove: StoredPanel, newPosition: LayoutUpdate) => {
+      return true // for now, just enable all drop areas while we are tweaking the behavior
+      const wouldBePanelState = updateLayout(panelState, itemToMove, newPosition)
+      const wouldBePanelStateEqualsCurrentPanelState = panelState.every((column, colIndex) =>
+        column.every(
+          (item, itemIndex) => item.name === wouldBePanelState[colIndex]?.[itemIndex]?.name,
+        ),
+      )
+
+      if (wouldBePanelStateEqualsCurrentPanelState) {
+        // if the drop results in no change, we don't allow it
+        return false
+      }
+
+      return true
+    },
+    [panelState],
+  )
+
+  const columnWidths: Array<`${number}px`> = React.useMemo(
```

Comment:
This doesn't compile...

Reply:
This is actually valid TS!

Reply:
oh, old bundled typescript vs new typescript running in vscode tripped me.